### PR TITLE
multi exposure

### PR DESCRIPTION
**Pull request by Gustry**
### What does it fix?
* Ticket: #4437 
* Funded by: DFAT
* Description: Add the multi exposure analysis

screen shot 2017-09-27 at 12 44 10

- [ ] Update string help
- [ ] Return from the wrapper

### Checklist:
- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [ ] Request someone to review or test your PR

### CODE REVIEW

File: safe/gui/tools/multi_exposure_dialog.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import logging
+
+
+from PyQt4.QtGui import QDialog, QComboBox, QLabel
+from safe.definitions.exposure import exposure_all
+from safe.utilities.resources import (
+    get_ui_class,
+)
+from safe.utilities.i18n import tr
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [F401] 'safe.utilities.i18n.tr' imported but unused

File: safe/gui/tools/multi_exposure_dialog.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    def _create_exposure_combos(self):
+        """Create one combobox for each exposure and insert them in the UI."""
+        for one_exposure in exposure_all:
+            label = QLabel(one_exposure['name'])
+            combo = QComboBox()
+            self.frame_exposures.addWidget(label)
+            self.frame_exposures.addWidget(combo)
+
+
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [W391] blank line at end of file

File: safe/impact_function/multi_exposure_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# coding=utf-8
+
+"""
+Multi-exposure wrapper.
+
+This class will manage how to launch and optimize a multi exposure analysis.
+"""
+
+from safe.impact_function.impact_function import ImpactFunction
```
Comment (ismailsunni) [inline safe/impact_function/multi_exposure_wrapper.py]: __[flake8]__
* [F401] 'safe.impact_function.impact_function.ImpactFunction' imported but unused

File: safe/impact_function/test/test_multi_exposure_impact_function.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+# coding=utf-8
+
+"""Test for Multi Exposure Impact Function."""
+
+from qgis.testing import TestCase
```
Comment (ismailsunni) [inline safe/impact_function/test/test_multi_exposure_impact_function.py]: __[flake8]__
* [F401] 'qgis.testing.TestCase' imported but unused

File: safe/impact_function/test/test_multi_exposure_impact_function.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+        impact_function = MultiExposureImpactFunction()
+        impact_function.hazard = hazard_layer
+        impact_function.exposures = [
+            building_layer, population_layer, roads_layer]
+        impact_function.aggregation = aggregation_layer
+
+
+
+
```
Comment (ismailsunni) [inline safe/impact_function/test/test_multi_exposure_impact_function.py]: __[flake8]__
* [W391] blank line at end of file

File: safe/gui/tools/multi_exposure_dialog.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+from qgis.core import QgsMapLayerRegistry
+from qgis.utils import iface
+
+from safe.common.exceptions import (
+    NoKeywordsFoundError,
+    KeywordNotFoundError,
+    MetadataReadError,
+)
+from safe.definitions.constants import inasafe_keyword_version_key, PREPARE_MULTI_SUCCESS, ANALYSIS_MULTI_SUCCESS
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [F401] 'safe.definitions.constants.ANALYSIS_MULTI_SUCCESS' imported but unused
* [F401] 'safe.definitions.constants.PREPARE_MULTI_SUCCESS' imported but unused
* [E501] line too long (113 > 79 characters)

File: safe/gui/tools/multi_exposure_dialog.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+from safe.definitions.exposure import exposure_all
+from safe.definitions.layer_purposes import layer_purpose_hazard, layer_purpose_exposure, layer_purpose_aggregation
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [E501] line too long (115 > 79 characters)

File: safe/gui/tools/multi_exposure_dialog.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+from safe.gui.analysis_utilities import add_impact_layers_to_canvas
+from safe.gui.gui_utilities import layer_from_combo, add_ordered_combo_item
+from safe.impact_function.multi_exposure_wrapper import MultiExposureImpactFunction
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [E501] line too long (83 > 79 characters)

File: safe/gui/tools/multi_exposure_dialog.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        self.setupUi(self)
+        self.combos_exposures = {}
+        # import pydevd
+        # pydevd.settrace('localhost', port=8999, stdoutToServer=True,
+        #                 stderrToServer=True)
+        self.keyword_io = KeywordIO()
+        self._create_exposure_combos()
+
+        # Fix for issue 1699 - cancel button does nothing
+        cancel_button = self.button_box.button(QDialogButtonBox.Cancel)
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [F841] local variable 'cancel_button' is assigned to but never used

File: safe/gui/tools/multi_exposure_dialog.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        # self.assertEqual(code, PREPARE_MULTI_SUCCESS, message)
+
+        code, message = impact_function.run()
+        # self.assertEqual(code, ANALYSIS_MULTI_SUCCESS, message)
+
+        for analysis in impact_function.impact_functions():
+            add_impact_layers_to_canvas(self.impact_function, iface)
+
+
+
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [W391] blank line at end of file

File: safe/impact_function/multi_exposure_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+        self._impact_functions = []
+
+        # We delegate the prepare to the main IF for each exposure
+        for exposure in self._exposures:
+            impact_function = ImpactFunction()
+            impact_function.hazard = deep_duplicate_layer(self._hazard)
+            impact_function.exposure = exposure
+            if self._aggregation:
+                impact_function.aggregation = deep_duplicate_layer(self._aggregation)
```
Comment (ismailsunni) [inline safe/impact_function/multi_exposure_wrapper.py]: __[flake8]__
* [E501] line too long (85 > 79 characters)

File: safe/utilities/gis.py [hunk 12]
```diff
@@ -170,6 +171,25 @@ def validate_geo_array(extent):
 
     return True
 
+def deep_duplicate_layer(layer):
```
Comment (ismailsunni) [inline safe/utilities/gis.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1

File: safe/gui/tools/multi_exposure_dialog.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+        self.setupUi(self)
+        self.combos_exposures = {}
+        self.keyword_io = KeywordIO()
+        self._create_exposure_combos()
+        self._multi_exposure_if = None
+
+        enable_messaging(self.message_viewer, self)
+
+        # Fix for issue 1699 - cancel button does nothing
+        cancel_button = self.button_box.button(QDialogButtonBox.Cancel)
```
Comment (ismailsunni) [inline safe/gui/tools/multi_exposure_dialog.py]: __[flake8]__
* [F841] local variable 'cancel_button' is assigned to but never used

File: safe/impact_function/test/test_impact_function.py [hunk 14]
```diff
@@ -113,6 +113,7 @@
 from safe.utilities.gis import wkt_to_rectangle
 from safe.utilities.utilities import readable_os_version
 from safe.impact_function.impact_function import ImpactFunction
+from safe.impact_function.impact_function_utilities import check_input_layer
```
Comment (ismailsunni) [inline safe/impact_function/test/test_impact_function.py]: __[flake8]__
* [E402] module level import not at top of file

File: safe/utilities/gis.py [hunk 15]
```diff
@@ -10,7 +10,14 @@
     QgsCoordinateTransform,
     QGis,
     QgsRectangle,
+<<<<<<< HEAD
```
Comment (ismailsunni) [inline safe/utilities/gis.py]: __[flake8]__
* [E999] SyntaxError: invalid syntax
* [E225] missing whitespace around operator
* [E227] missing whitespace around bitwise or shift operator

File: safe/utilities/gis.py [hunk 16]
```diff
@@ -10,7 +10,14 @@
...
 )
+=======
```
Comment (ismailsunni) [inline safe/utilities/gis.py]: __[flake8]__
* [E225] missing whitespace around operator

File: safe/utilities/gis.py [hunk 17]
```diff
@@ -10,7 +10,14 @@
...
+    QgsPoint,
```
Comment (ismailsunni) [inline safe/utilities/gis.py]: __[flake8]__
* [E113] unexpected indentation

File: safe/utilities/gis.py [hunk 18]
```diff
@@ -10,7 +10,14 @@
...
+    QgsVectorLayer,
+    QgsRasterLayer)
+from safe.utilities.metadata import copy_layer_keywords
+>>>>>>> multi exposure
```
Comment (ismailsunni) [inline safe/utilities/gis.py]: __[flake8]__
* [W503] line break before binary operator
* [E225] missing whitespace around operator

File: safe/impact_function/multi_exposure_wrapper.py [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+    def impact_functions(self):
+        """Return the list of impact functions which have been used.
+
+        :return: List of impact functions.
+        :rtype: list(ImpactFunction)
+        """
+        return self._impact_functions
+
+    @property
+    def if_being_processed(self):
```
Comment (ismailsunni) [inline safe/impact_function/multi_exposure_wrapper.py]: The first time I read, I thought it's like a checker/validation method. Perhaps change it to more easy to read (e.g. `current_impact_function`, or another name :D )

Comment (Gustry) [reply safe/impact_function/multi_exposure_wrapper.py]: Good catch, you're right.

**Pull request merged**